samza-core/src/main/java/org/apache/samza/storage/TransactionalTaskSideInputStorageManager.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.samza.util.Clock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class TransactionalTaskSideInputStorageManager extends NonTransactionalTaskSideInputStorageManager implements TaskSideInputStorageManager {
+  private static final Logger LOG = LoggerFactory.getLogger(TransactionalTaskSideInputStorageManager.class);
+
+  public TransactionalTaskSideInputStorageManager(
+      TaskName taskName,
+      TaskMode taskMode,
+      File storeBaseDir,
+      Map<String, StorageEngine> sideInputStores,
+      Map<String, Set<SystemStreamPartition>> storesToSSPs,
+      Clock clock) {
+    super(taskName, taskMode, storeBaseDir, sideInputStores, storesToSSPs, clock);
+  }
+
+  @Override
+  public Map<String, Path> checkpoint(CheckpointId checkpointId) {
+    LOG.info("Creating checkpoint for task: {}", this.taskName);
+
+    Map<String, Path> checkpointPaths = new HashMap<>();
+    stores.forEach((store, storageEngine) ->
+      // TODO what subset of stores to checkpoint? an ssp can be a changelog side input for one store and a regular side input for another store
+      storageEngine.checkpoint(checkpointId).ifPresent(path -> checkpointPaths.put(store, path))
+    );
+    return checkpointPaths;
+  }
+
+  @Override
+  public void removeOldCheckpoints(String latestCheckpointId) {
+    LOG.info("Removing checkpoints older than: {} for task: {}", latestCheckpointId, this.taskName);
+    File[] storeDirs = storeBaseDir.listFiles((dir, name) -> stores.containsKey(name));
+    (storeDirs == null ? Stream.<File>empty() : Arrays.stream(storeDirs)).forEach(storeDir -> {
+        String taskStoreName = storageManagerUtil.getTaskStoreDir(storeBaseDir, storeDir.getName(), taskName, taskMode).getName();
+        FileFilter wildcardFileFilter = new WildcardFileFilter(taskStoreName + "-*");
+        File[] checkpointDirs = storeDir.listFiles(wildcardFileFilter);
+        (checkpointDirs == null ? Stream.<File>empty() : Arrays.stream(checkpointDirs)).forEach(checkpointDir -> {
+            if (checkpointDir.getName().contains(latestCheckpointId)) {
+              try {
+                FileUtils.deleteDirectory(checkpointDir);
+              } catch (IOException e) {
+                LOG.error("Failed to remove old checkpointDir: {} for task: {}, latestCheckpointId: {}", checkpointDir, taskName, latestCheckpointId);
+              }
+            }
+          });
+      });
+  }
+
+  /**
+   * This method will write offset files to each store's primary and checkpoint directories (if given) corresponding
+   * to the set of lastProcessedOffsets given.
+   */
+  @Override
+  public void writeOffsetFiles(Map<SystemStreamPartition, String> lastProcessedOffsets, Map<String, Path> checkpointPaths) {
+    // write offset files for each store's non-checkpoint directories
+    super.writeOffsetFiles(lastProcessedOffsets, Collections.emptyMap());
+
+    // write offset files to checkpoint directories
```

Comment:
Side input consumption on restart will resume from the last marked offset in the primary store directory, and at the next commit it will create a new checkpoint directory with a new id, and clean up any previous checkpoint directories including the one which had no offset file.

In the case that a failover to this container happens before it can be restarted and (possibly) have its store directory contents remedied as above, transactional restore will ignore the bad checkpoint directory use the previous one. There should be another, older slightly directory, as failing to write offset directories will throw an exception and prevent `removeOldCheckpoints` from cleaning it up. Restore will use this directory and load anything between its offset and the checkpointed offset from changelog. If there is no previous checkpoint directory, either we've never successfully created a checkpoint indicating a bigger problem, or the active task has not yet checkpointed anything, preventing the standby task from creating one.

Does this make sense? Happy to chat more about this if needed.